fpga_interchange/arch_pack_clusters.cc

```diff
@@ -0,0 +1,496 @@
+
+                    if (bel_data.category != BEL_CATEGORY_LOGIC)
+                        continue;
+
+                    if (bel_data.synthetic)
+                        continue;
+
+                    if (ctx->getBelPinType(bel_pin.bel, bel_pin.pin) != PORT_IN)
+                        continue;
+
+                    if (sink_bel.index == bel.index)
+                        return sink_bel;
+                }
+            }
+        }
+    }
+
+    BelId tmp_bel;
+    return tmp_bel;
+}
+
+CellInfo* Arch::getClusterRootCell(ClusterId cluster) const
+{
+    NPNR_ASSERT(cluster != ClusterId());
+    CellInfo *ci = nullptr;
+    auto cell = getCtx()->cells.find(cluster);
+    if (cell != getCtx()->cells.end()) {
+        ci = cell->second.get();
+    }
+    return ci;
+}
+
+bool Arch::getClusterPlacement(ClusterId cluster, BelId root_bel,
+                         std::vector<std::pair<CellInfo *, BelId>> &placement) const
+{
+    const Context *ctx = getCtx();
+
+    // Place root
+    CellInfo *root_cell = getClusterRootCell(cluster);
+
+    if (!ctx->isValidBelForCellType(root_cell->type, root_bel))
+        return false;
+
+    // Place cluster
+    // Prepare first loop iteration
+    CellInfo *next_cell = root_cell;
+    NetInfo *next_net = nullptr;
+    BelId next_bel = root_bel;
+    Loc next_loc = getBelLocation(next_bel);
+
+    std::vector<IdString> placed_bels;
+
+    do {
+        //log_info("Cell name: %s - Cell type: %s - Bel: %s\n", next_cell->name.c_str(ctx), next_cell->type.c_str(ctx), ctx->nameOfBel(next_bel));
+        placement.push_back(std::make_pair(next_cell, next_bel));
+
+        // Place driver cells specified according to the BEL chain information
+        placed_bels.clear();
+        auto &next_bel_data = bel_info(ctx->chip_info, next_bel);
+        for (auto port : cluster_port_bels.at(cluster)) {
+            IdString driver_port(port.first);
+
+            CellInfo *driver_cell = next_cell->ports.at(driver_port).net->driver.cell;
+
+            bool cell_ok = false;
+            for (auto &cell : cluster_driver_cells.at(cluster)) {
+                if (IdString(cell) == driver_cell->type) {
+                    cell_ok = true;
+                    break;
+                }
+            }
+
+            for (auto driver_in_port : driver_cell->ports) {
+                if (cell_ok)
+                    break;
+
+                if (driver_in_port.second.type != PORT_IN)
+                    continue;
+
+                CellInfo *driver_cell_driver = driver_in_port.second.net->driver.cell;
+
+                for (auto &cell : cluster_driver_in_out_cells.at(cluster)) {
+                    if (IdString(cell) == driver_cell->type && driver_cell_driver == next_cell) {
+                        cell_ok = true;
+                        break;
+                    }
+                }
+            }
+
+            if (!cell_ok or driver_cell == nullptr)
```

Comment:
please use `||` not `or`

Reply:
Sure, I guess I just came back from python when writing this :)